runtime/jcl/common/shared.c

```diff
@@ -563,23 +563,32 @@ getCpeTypeForProtocol(char* protocol, jsize protocolLen, const char* pathChars,
 	}
 	if (strncmp(protocol, "file", 5)==0) {
 		char* endsWith = (char*)(pathChars + (pathLen-4));
-		if ((strncmp(endsWith, ".jar", 4)==0) || (strncmp(endsWith, ".zip", 4)==0) || strstr(pathChars,"!/") || strstr(pathChars,"!\\")) {
-			Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_ExitJAR();
-			return CPE_TYPE_JAR;
-		} else {
-			char JimageEndsWith[] = DIR_SEPARATOR_STR "lib" DIR_SEPARATOR_STR "modules";
-			IDATA len = LITERAL_STRLEN(JimageEndsWith);
-
-			if (pathLen >= len) {
-				endsWith = (char*)(pathChars + (pathLen - len));
-				if (strncmp(endsWith, JimageEndsWith, len) == 0) {
-					Trc_JCL_com_ibm_oti_shared_getCpeTypeForProtocol_ExitJIMAGE();
-					return CPE_TYPE_JIMAGE;
+		int32_t fileType;
+		PORT_ACCESS_FROM_JAVAVM(vm);
+		fileType = j9file_attr(pathChars);
```

Comment:
`j9file_attr()` could have some overhead. We can optimize the code a bit so that we do not have to call it in every case. For jimage and fat jar (which has `!/` or `!\\`), we know they are files. Probably something like:

```
if (strncmp(protocol, "file", 5)==0) {
	if ((strstr(pathChars,"!/") || strstr(pathChars,"!\\")) {
		return CPE_TYPE_JAR;
	} else if (isJimage()) {
                /* You may need to add helper isJimage() */
		return CPE_TYPE_JIMAGE;
	} else {
		if (EsIsFile == j9file_attr(pathChars)) {
			return CPE_TYPE_JAR;
		} else {
			return CPE_TYPE_DIRECTORY;
		}
	}
}
```

Reply:
Updated.